File: quantum_yolo_engine/validate.py
# quantum_yolo_engine/validate.py
from __future__ import annotations

import math
from typing import Dict, List

from .models import PRODUCT_IDS, AssetStrategy


class StrategyValidationError(ValueError):
    """Raised with every violation found, not just the first, so callers can
    surface a complete list of problems in one pass."""

    def __init__(self, errors: List[str]):
        self.errors = errors
        super().__init__("; ".join(errors))


def _finite(x: float) -> bool:
    try:
        return math.isfinite(float(x))
    except (TypeError, ValueError):
        return False


def validate_strategy(bankroll_usd: float, strategies: Dict[str, AssetStrategy]) -> None:
    """Validates a full strategy (bankroll + all assets). Raises
    StrategyValidationError containing every violation found."""
    errors: List[str] = []

    if not _finite(bankroll_usd) or bankroll_usd <= 0:
        errors.append(f"bankroll_usd must be a positive finite number, got {bankroll_usd!r}")

    enabled_allocation_total = 0.0

    for product_id, strat in strategies.items():
        prefix = f"{product_id}: "

        if product_id not in PRODUCT_IDS:
            errors.append(f"{prefix}unsupported product_id (supported: {PRODUCT_IDS})")

        if not _finite(strat.allocation_usd) or strat.allocation_usd <= 0:
            errors.append(f"{prefix}allocation_usd must be a positive finite number, got {strat.allocation_usd!r}")
        elif strat.enabled:
            enabled_allocation_total += strat.allocation_usd

        if not _finite(strat.stop_price) or strat.stop_price <= 0:
            errors.append(f"{prefix}stop_price must be a positive finite number, got {strat.stop_price!r}")

        tp = strat.take_profit
        if not _finite(tp.tp1_price) or tp.tp1_price <= 0:
            errors.append(f"{prefix}take_profit.tp1_price must be a positive finite number")
        if not _finite(tp.tp2_price) or tp.tp2_price <= 0:
            errors.append(f"{prefix}take_profit.tp2_price must be a positive finite number")
        if not _finite(tp.tp1_fraction) or not (0 < tp.tp1_fraction <= 1):
            errors.append(f"{prefix}take_profit.tp1_fraction must satisfy 0 < fraction <= 1, got {tp.tp1_fraction!r}")
        if not _finite(tp.tp2_fraction) or not (0 < tp.tp2_fraction <= 1):
            errors.append(f"{prefix}take_profit.tp2_fraction must satisfy 0 < fraction <= 1, got {tp.tp2_fraction!r}")
        if _finite(tp.tp1_fraction) and _finite(tp.tp2_fraction):
            total_exit = tp.tp1_fraction + tp.tp2_fraction
            if total_exit > 1.0 + 1e-9:
                errors.append(
                    f"{prefix}tp1_fraction + tp2_fraction must be <= 1.0, got {total_exit!r}"
                )

        if (
            _finite(tp.tp1_price)
            and _finite(tp.tp2_price)
            and tp.tp1_price > 0
            and tp.tp2_price > 0
            and tp.tp1_price >= tp.tp2_price
        ):
            errors.append(f"{prefix}take_profit.tp1_price must be < tp2_price")

        if (
            _finite(strat.stop_price)
            and _finite(tp.tp1_price)
            and strat.stop_price > 0
            and tp.tp1_price > 0
            and strat.stop_price >= tp.tp1_price
        ):
            errors.append(f"{prefix}stop_price must be < take_profit.tp1_price")

        seen_ids = set()
        if not strat.entries:
            errors.append(f"{prefix}must have at least one entry rule")

        for entry in strat.entries:
            if entry.id in seen_ids:
                errors.append(f"{prefix}duplicate entry id {entry.id!r}")
            seen_ids.add(entry.id)

            if not _finite(entry.price) or entry.price <= 0:
                errors.append(f"{prefix}entry {entry.id!r} price must be a positive finite number")
            elif _finite(strat.stop_price) and strat.stop_price > 0 and entry.price <= strat.stop_price:
                errors.append(f"{prefix}entry {entry.id!r} price must be > stop_price")

            if not _finite(entry.quote_size_usd) or entry.quote_size_usd <= 0:
                errors.append(f"{prefix}entry {entry.id!r} quote_size_usd must be a positive finite number")

        if strat.enabled and strat.entries:
            entries_total = sum(e.quote_size_usd for e in strat.entries if _finite(e.quote_size_usd))
            if _finite(strat.allocation_usd) and entries_total > strat.allocation_usd + 1e-9:
                errors.append(
                    f"{prefix}entries total (${entries_total:.2f}) exceeds allocation_usd "
                    f"(${strat.allocation_usd:.2f})"
                )

    if _finite(bankroll_usd) and enabled_allocation_total > bankroll_usd + 1e-9:
        errors.append(
            f"total enabled allocations (${enabled_allocation_total:.2f}) exceed bankroll_usd (${bankroll_usd:.2f})"
        )

    if errors:
        raise StrategyValidationError(errors)

File: web/src/engine/validate.ts
import { PRODUCT_IDS, type AssetStrategy, type Strategy } from "./model";

export class StrategyValidationError extends Error {
  readonly errors: string[];
  constructor(errors: string[]) {
    super(errors.join("; "));
    this.name = "StrategyValidationError";
    this.errors = errors;
  }
}

export interface FieldError {
  path: string;
  message: string;
}

function isFinitePositive(x: number): boolean {
  return Number.isFinite(x) && x > 0;
}

/**
 * Validates a full strategy. Returns every violation found (never throws),
 * so the UI can attach each error to its field. Callers that want the
 * throw-on-invalid behavior (e.g. before starting a run) should check
 * `errors.length === 0` themselves or use `assertValidStrategy`.
 */
export function validateStrategy(strategy: Strategy): FieldError[] {
  const errors: FieldError[] = [];

  if (!isFinitePositive(strategy.bankrollUsd)) {
    errors.push({ path: "bankrollUsd", message: "must be a positive finite number" });
  }

  if (!strategy.assets.length) {
    errors.push({ path: "assets", message: "must have at least one asset" });
  }

  let enabledAllocationTotal = 0;
  const seenProductIds = new Set<string>();

  strategy.assets.forEach((asset, assetIndex) => {
    const base = `assets[${assetIndex}]`;

    if (seenProductIds.has(asset.productId)) {
      errors.push({ path: `${base}.productId`, message: `duplicate product ${asset.productId}` });
    }
    seenProductIds.add(asset.productId);

    if (!(PRODUCT_IDS as readonly string[]).includes(asset.productId)) {
      errors.push({ path: `${base}.productId`, message: `unsupported product_id (supported: ${PRODUCT_IDS.join(", ")})` });
    }

    if (!isFinitePositive(asset.allocationUsd)) {
      errors.push({ path: `${base}.allocationUsd`, message: "must be a positive finite number" });
    } else if (asset.enabled) {
      enabledAllocationTotal += asset.allocationUsd;
    }

    if (!isFinitePositive(asset.stopPrice)) {
      errors.push({ path: `${base}.stopPrice`, message: "must be a positive finite number" });
    }

    const tp = asset.takeProfit;
    if (!isFinitePositive(tp.tp1Price)) {
      errors.push({ path: `${base}.takeProfit.tp1Price`, message: "must be a positive finite number" });
    }
    if (!isFinitePositive(tp.tp2Price)) {
      errors.push({ path: `${base}.takeProfit.tp2Price`, message: "must be a positive finite number" });
    }
    if (!Number.isFinite(tp.tp1Fraction) || !(tp.tp1Fraction > 0 && tp.tp1Fraction <= 1)) {
      errors.push({ path: `${base}.takeProfit.tp1Fraction`, message: "must satisfy 0 < fraction <= 1" });
    }
    if (!Number.isFinite(tp.tp2Fraction) || !(tp.tp2Fraction > 0 && tp.tp2Fraction <= 1)) {
      errors.push({ path: `${base}.takeProfit.tp2Fraction`, message: "must satisfy 0 < fraction <= 1" });
    }
    if (Number.isFinite(tp.tp1Fraction) && Number.isFinite(tp.tp2Fraction)) {
      const total = tp.tp1Fraction + tp.tp2Fraction;
      if (total > 1 + 1e-9) {
        errors.push({ path: `${base}.takeProfit`, message: "tp1Fraction + tp2Fraction must be <= 1.0" });
      }
    }
    if (isFinitePositive(tp.tp1Price) && isFinitePositive(tp.tp2Price) && tp.tp1Price >= tp.tp2Price) {
      errors.push({ path: `${base}.takeProfit.tp1Price`, message: "must be < tp2Price" });
    }
    if (isFinitePositive(asset.stopPrice) && isFinitePositive(tp.tp1Price) && asset.stopPrice >= tp.tp1Price) {
      errors.push({ path: `${base}.stopPrice`, message: "must be < takeProfit.tp1Price" });
    }

    if (!asset.entries.length) {
      errors.push({ path: `${base}.entries`, message: "must have at least one entry rule" });
    }

    const seenEntryIds = new Set<string>();
    let entriesTotal = 0;
    asset.entries.forEach((entry, entryIndex) => {
      const entryBase = `${base}.entries[${entryIndex}]`;
      if (seenEntryIds.has(entry.id)) {
        errors.push({ path: `${entryBase}.id`, message: `duplicate entry id ${entry.id}` });
      }
      seenEntryIds.add(entry.id);

      if (!isFinitePositive(entry.price)) {
        errors.push({ path: `${entryBase}.price`, message: "must be a positive finite number" });
      } else if (isFinitePositive(asset.stopPrice) && entry.price <= asset.stopPrice) {
        errors.push({ path: `${entryBase}.price`, message: "must be > stopPrice" });
      }

      if (!isFinitePositive(entry.quoteSizeUsd)) {
        errors.push({ path: `${entryBase}.quoteSizeUsd`, message: "must be a positive finite number" });
      } else {
        entriesTotal += entry.quoteSizeUsd;
      }
    });

    if (asset.enabled && isFinitePositive(asset.allocationUsd) && entriesTotal > asset.allocationUsd + 1e-9) {
      errors.push({
        path: `${base}.entries`,
        message: `entries total ($${entriesTotal.toFixed(2)}) exceeds allocationUsd ($${asset.allocationUsd.toFixed(2)})`,
      });
    }
  });

  if (isFinitePositive(strategy.bankrollUsd) && enabledAllocationTotal > strategy.bankrollUsd + 1e-9) {
    errors.push({
      path: "bankrollUsd",
      message: `total enabled allocations ($${enabledAllocationTotal.toFixed(2)}) exceed bankrollUsd ($${strategy.bankrollUsd.toFixed(2)})`,
    });
  }

  return errors;
}

export function assertValidStrategy(strategy: Strategy): void {
  const errors = validateStrategy(strategy);
  if (errors.length) {
    throw new StrategyValidationError(errors.map((e) => `${e.path}: ${e.message}`));
  }
}

export function isValidAssetStrategy(asset: AssetStrategy): boolean {
  return (PRODUCT_IDS as readonly string[]).includes(asset.productId);
}
